### ISSUE TITLE
Self assignment at declaration (ambiguous construct)

### ISSUE THREAD

**Issue by axic** [ethereum/solidity#2495]
The following compiles without issues:

```
contract C {
  function f() {
    int a = a;
  }
}
```

Since variables are pre initialised to 0, the above should be effectively a no-op. I do think however it can be confusing. I would suggest throwing an error in this case or at a bare minimum a warning.

Think about:
```
  int aa = 5;
  int a = 4 * a; // user intended to use aa
```

The same applies to `var`, though that is already rejected (for a different reason):
```
2988.sol:3:13: Error: Declaration referenced before type could be determined.
    var a = a;
            ^
```

**Comment by axic** [comment]
@pirapira @chriseth comments?

**Comment by pirapira** [comment]
What about a mutually referencing initialization
```
int aa = 5 * a + 1;
int a = aa - 5;
```
does not do
`(a, aa) = (1, 6)`.

**Comment by axic** [comment]
I think that compiles to:
```
int aa = 5 * 0 + 1;
int a = aa - 5;
```

e.g. -4

**Comment by pirapira** [comment]
So, the `a` in the first line `int aa = 5 * a + 1` could be `aa` or `int b` or whatever not-yet-initialized variable.  I opt for making not-yet-initialized variables out of scope because they are so meaningless that they are mutually replaceable as long as types match.

**Comment by federicobond** [comment]
This is a remnant of the JavaScript style scoping. I thought there were no plans to change it, but maybe we have an open window to make such a change in time for 0.5.0. I don't think many users are relying on it, and it's definitely not a style we would want to encourage.

**Comment by axic** [comment]
I am on my mission of getting rid of that :)

**Comment by axic** [comment]
To be fixed by #1679.

**Comment by axic** [comment]
Fixed by #3476.
